### PR TITLE
Implement UnionPhase & UnionPlan

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

**Comment by matriv** [comment]
@seut Please review

File: sql/src/main/java/io/crate/jobs/UnionContext.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    public int id() {
+        return unionPhase.executionPhaseId();
+    }
+
+    @Override
+    protected void innerPrepare() throws Exception {
+        flatProjectorChain.prepare();
+    }
+
+//    @Override
```
Comment (seut) [inline sql/src/main/java/io/crate/jobs/UnionContext.java]: remove commented lines

File: sql/src/main/java/io/crate/action/job/ContextPreparer.java [hunk 2]
```diff
@@ -608,6 +603,69 @@ public Boolean visitNestedLoopPhase(NestedLoopPhase phase, PreparerContext conte
...
+
+            MultiUpstreamRowReceiver multiUpstreamRowReceiver =
+                new MultiUpstreamRowReceiver(flatProjectorChain.firstProjector());
+
+            Map<Byte, PageDownstreamContext> downstreamContexts = new HashMap<>(phase.mergePhases().size());
+            short inputId = 0;
+
+            for (MergePhase mergePhase : phase.mergePhases()) {
+                // TODO: Replace with this once new FetchProjector is ready
+//              RowReceiver rowReceiver = new PrependInputIdRowReceiver(multiUpstreamRowReceiver.newRowReceiver(), (byte) inputId),
```
Comment (seut) [inline sql/src/main/java/io/crate/action/job/ContextPreparer.java]: as discussed, this must always work, not related to the fetch projector

File: sql/src/main/java/io/crate/planner/node/dql/MergePhase.java [hunk 3]
```diff
@@ -137,6 +139,46 @@ public static MergePhase sortedMerge(UUID jobId,
         return mergeNode;
     }
 
+    // TODO: this is a duplicate, it coexists in QueryThenFetchConsumer
```
Comment (seut) [inline sql/src/main/java/io/crate/planner/node/dql/MergePhase.java]: could you maybe check and eliminate the duplicate?

Comment (seut) [inline sql/src/main/java/io/crate/planner/node/dql/MergePhase.java]: add @Nullable

File: sql/src/main/java/io/crate/planner/node/dql/UnionPhase.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    public UnionPhase(UUID jobId,
+                      int executionNodeId,
+                      String name,
+                      List<Projection> projections,
+                      Collection<MergePhase> mergePhases,
+                      Collection<String> executionNodes) {
+        super(jobId, executionNodeId, name, projections);
+        assert !mergePhases.isEmpty();
+        if (projections.isEmpty()) {
+            outputTypes = Iterables.getLast(mergePhases).outputTypes();
```
Comment (seut) [inline sql/src/main/java/io/crate/planner/node/dql/UnionPhase.java]: last one could be null. anyway it looks a bit odd as due to lastOne usage it looks a bit like mergePhases are chained, I'd suggest add at least a comment that it doesn't matter which mergePhase is used (as long as it is not null)

File: sql/src/main/java/io/crate/planner/node/dql/UnionPhase.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        return mergePhases;
+    }
+
+    @Override
+    public <C, R> R accept(ExecutionPhaseVisitor<C, R> visitor, C context) {
+        return visitor.visitUnionPhase(this, context);
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
```
Comment (seut) [inline sql/src/main/java/io/crate/planner/node/dql/UnionPhase.java]: UnionPhase currently should never be streamed, throw an exception? (same at writeTo)

File: sql/src/test/java/io/crate/integrationtests/UnionSingleNodeIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package io.crate.integrationtests;
+
+import io.crate.testing.UseJdbc;
+import org.elasticsearch.test.ESIntegTestCase;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 1, numClientNodes = 0)
+@UseJdbc
+public class UnionSingleNodeIntegrationTest extends UnionIntegrationTest {
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/UnionSingleNodeIntegrationTest.java]: afaik this class is not needed as union phases are currently never distributed, remove.

Comment (matriv) [inline sql/src/test/java/io/crate/integrationtests/UnionSingleNodeIntegrationTest.java]: We'll keep it since with 1 node we skip the intermediate merge phases and it's good to test, I'm changing the comment to describe that.

File: sql/src/main/java/io/crate/planner/node/dql/MergePhase.java (hunk 3 above)
Comment (matriv) [inline sql/src/main/java/io/crate/planner/node/dql/MergePhase.java]: Will do in a another PR

File: core/src/main/java/io/crate/core/collections/PrependedRow.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Override
+    public Object get(int index) {
+        return index == 0 ? inputId : delegateRow.get(index - 1);
+    }
+
+    @Override
+    public Object[] materialize() {
+        return delegateRow.materialize();
```
Comment (mfussenegger) [inline core/src/main/java/io/crate/core/collections/PrependedRow.java]: Seems odd that materialize doesn't include the inputId.
This means `Bucket.materialize(row)` will have a different result than `row.materialize()`. If this is required behaviour for some reason this should be documented

File: sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java [hunk 8]
```diff
@@ -119,6 +119,14 @@ public AnalyzedRelation visitMultiSourceSelect(MultiSourceSelect multiSourceSele
     public AnalyzedRelation visitTwoRelationsUnion(TwoRelationsUnion twoTableUnion, Context context) {
         process(twoTableUnion.first(), context);
         process(twoTableUnion.second(), context);
+
+        // PushDown  limit
+        Optional<Symbol> pushDownLimit = Limits.add(twoTableUnion.querySpec().limit(),
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java]: can you extend the comment to explain more on a semantic level what is happening here? why limit + offset?

File: sql/src/main/java/io/crate/planner/node/dql/MergePhase.java (hunk 3 above)
Comment (matriv) [inline sql/src/main/java/io/crate/planner/node/dql/MergePhase.java]: Will do in a different PR
